### PR TITLE
LSIF: GraphQL endpoints for LSIF jobs

### PR DESCRIPTION
**Pull request by efritz**
This PR adds the GraphQL endpoints for LSIF jobs as defined in [RFC 46](https://docs.google.com/document/d/1v1Sojj4B07ZTX3bI3pu6uVCRGkYhujkQXmIi6A3Nbc4).

### CODE REVIEW

File: cmd/frontend/graphqlbackend/lsif_jobs.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+		var path string
+		if r.opt.NextURL == nil {
+			// first page of results
+			path = fmt.Sprintf("/jobs/%s", strings.ToLower(r.opt.Status))
+		} else {
+			// subsequent page of results
+			path = *r.opt.NextURL
+		}
+
+		fmt.Printf("> %#v\n", path)
```
Comment (chrismwendt) [inline cmd/frontend/graphqlbackend/lsif_jobs.go]: Leftover from debugging?

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 2]
```diff
@@ -1293,6 +1293,34 @@ type Query {
         # 'LSIFDumpConnection.pageInfo.endCursor' that is returned.
         after: String
     ): LSIFDumpConnection!
+
+    # Retrieve counts of jobs in the LSIF work queue.
```
Comment (chrismwendt) [inline cmd/frontend/graphqlbackend/schema.graphql]:
```suggestion
    # Retrieve counts of jobs by status in the LSIF work queue.
```

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 3]
```diff
@@ -3927,6 +3955,81 @@ type LSIFDumpConnection {
...
+
+    # The number of completed jobs.
+    completed: Int!
+
+    # The number of failed jobs.
+    failed: Int!
+}
+
+# An LSIF job's status.
+enum LSIFJobState {
```
Comment (chrismwendt) [inline cmd/frontend/graphqlbackend/schema.graphql]: There are two terms here that mean almost the same thing: status/state. Any preference?

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 4]
```diff
@@ -1293,6 +1293,34 @@ type Query {
...
+
+    # Look up an LSIF job by ID.
+    lsifJob(id: ID!): LSIFJob
+
+    # Search for LSIF jobs by status ad query term.
+    lsifJobs(
+        # The status of returned jobs.
+        status: LSIFJobState!
+
+        # An (optional) search query taht searches over the job name, failure
```
Comment (chrismwendt) [inline cmd/frontend/graphqlbackend/schema.graphql]:
```suggestion
        # An (optional) search query that searches over the job name, failure
```

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 5]
```diff
@@ -3927,6 +3955,81 @@ type LSIFDumpConnection {
...
+    failed: Int!
+}
+
+# An LSIF job's status.
+enum LSIFJobState {
+    ACTIVE
+    QUEUED
+    SCHEDULED
+    COMPLETED
+    FAILED
```
Comment (chrismwendt) [inline cmd/frontend/graphqlbackend/schema.graphql]: Enums can be documented, too: https://github.com/sourcegraph/sourcegraph/blob/c42d03d2a5f463f9abc8c1290c4b250c176d15ef/cmd/frontend/graphqlbackend/schema.graphql#L500

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 6]
```diff
@@ -3927,6 +3955,81 @@ type LSIFDumpConnection {
...
+    id: ID!
+
+    # The job type (convert, or clean-old-jobs).
+    name: String!
+
+    # The job's arguments.
+    args: JSONValue!
+
+    # The job's current status.
+    status: LSIFJobState!
```
Comment (chrismwendt) [inline cmd/frontend/graphqlbackend/schema.graphql]: There's another example of state/status here: https://github.com/sourcegraph/sourcegraph/blob/c42d03d2a5f463f9abc8c1290c4b250c176d15ef/cmd/frontend/graphqlbackend/schema.graphql#L499-L522 Is there a similar distinction to be made here?

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 7]
```diff
@@ -3927,6 +3955,81 @@ type LSIFDumpConnection {
...
+
+    # The current job progress (0 to 100).
+    progress: Float!
+
+    # If the job failed, its failure message.
+    failedReason: String
+
+    # If the job failed, the its stacktrace.
```
Comment (chrismwendt) [inline cmd/frontend/graphqlbackend/schema.graphql]:
```suggestion
    # If the job failed, its stacktrace.
```

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 8]
```diff
@@ -3927,6 +3955,81 @@ type LSIFDumpConnection {
...
+    finishedOn: DateTime
+}
+
+# A list of LSIF jobs.
+type LSIFJobConnection {
+    # A list of LSIF jobs.
+    nodes: [LSIFJob!]!
+
+    # The total number of jobs with this status.
+    totalCount: Int!
```
Comment (chrismwendt) [inline cmd/frontend/graphqlbackend/schema.graphql]: Do all jobs in a given connection really have the same status?

File: cmd/frontend/graphqlbackend/lsif_jobs.go (hunk 1 above)
Comment (efritz) [reply cmd/frontend/graphqlbackend/lsif_jobs.go]: No I use a debugger like a real dev >_>

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 3 above)
Comment (efritz) [reply cmd/frontend/graphqlbackend/schema.graphql]: Status - I'll make sure they're consistent.

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 8 above)
Comment (efritz) [reply cmd/frontend/graphqlbackend/schema.graphql]: The only way to get a connection is to supply a status. The doc comment isn't correct but may seem odd here.

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 6 above)
Comment (efritz) [reply cmd/frontend/graphqlbackend/schema.graphql]: Maybe we should conform and choose state for the frontend? We could also use these similar names (with two additional). Thoughts?

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 8 above)
Comment (efritz) [reply cmd/frontend/graphqlbackend/schema.graphql]: I've updated this comment.

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 6 above)
Comment (efritz) [reply cmd/frontend/graphqlbackend/schema.graphql]: I've renamed Status to State to make it consistent with the type that exists.

File: cmd/frontend/graphqlbackend/lsif_job.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+	lsifJob *types.LSIFJob
+}
+
+func lsifJobByGQLID(ctx context.Context, id graphql.ID) (*lsifJobResolver, error) {
+	jobID, err := unmarshalLSIFJobGQLID(id)
+	if err != nil {
+		return nil, err
+	}
+
+	path := fmt.Sprintf("/jobs/%s", jobID)
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/lsif_job.go]: security: `jobID` comes from unmarshaling the GQL ID, so anyone could forge it. By using just string concatenation here, I could change this URL to `/jobs/../delete-job` or something perhaps and cause you to make that request below

Fix is to just make sure you are using `url.PathEscape`

File: cmd/frontend/graphqlbackend/lsif_job_stats.go [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+	"context"
+	"fmt"
+
+	graphql "github.com/graph-gophers/graphql-go"
+	"github.com/graph-gophers/graphql-go/relay"
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/internal/lsif"
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
+)
+
+const singletonLSIFJobStatsGQLID = "lsifJobStats"
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/lsif_job_stats.go]: consts in my mind don't match the typical connotation of a singleton / isn't worth calling out (isn't every global const a singleton?); I would name this just `lsifJobStatsGQLID`

File: cmd/frontend/graphqlbackend/lsif_job.go (hunk 9 above)
Comment (efritz) [reply cmd/frontend/graphqlbackend/lsif_job.go]: Good catch. Thanks!

File: cmd/frontend/graphqlbackend/lsif_job_stats.go (hunk 10 above)
Comment (efritz) [reply cmd/frontend/graphqlbackend/lsif_job_stats.go]: Sure. I was just copying what I saw from site and default settings.

**Pull request merged**